crates/ruff_python_formatter/src/expression/expr_tuple.rs

```diff
@@ -1,12 +1,98 @@
-use crate::{verbatim_text, FormatNodeRule, PyFormatter};
-use ruff_formatter::{write, Buffer, FormatResult};
+use crate::context::{NodeLevel, PyFormatContext};
+use crate::{AsFormat, FormatNodeRule, PyFormatter};
+use ruff_formatter::formatter::Formatter;
+use ruff_formatter::prelude::{
+    group, if_group_breaks, soft_block_indent, soft_line_break_or_space, text,
+};
+use ruff_formatter::{format_args, write, Buffer, Format, FormatResult};
+use ruff_python_ast::prelude::{Expr, Ranged};
+use ruff_text_size::{TextLen, TextRange};
 use rustpython_parser::ast::ExprTuple;
 
 #[derive(Default)]
 pub struct FormatExprTuple;
 
 impl FormatNodeRule<ExprTuple> for FormatExprTuple {
     fn fmt_fields(&self, item: &ExprTuple, f: &mut PyFormatter) -> FormatResult<()> {
-        write!(f, [verbatim_text(item.range)])
+        let ExprTuple {
+            range,
+            elts,
+            ctx: _,
+        } = item;
+
+        let saved_level = f.context().node_level();
```

Comment:
I believe the Expr formatting should already set the right node level for you

Reply:
i don't think it can, ExprTuple needs to know if it's outermost and Expr setting this would hide this information